### DISCUSSION TITLE
Can someone share a simple example on how to add vectors from CompreFace to milvus and search most similar vectors (faces)?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18132] [Q&A and General discussion] [answered]
Can someone share a simple example on how to add vectors from CompreFace to milvus and search most similar vectors (faces)?
I suppose you should use euclidian distance.

**Comment by yhmo** [comment, marked as answer]
I think it is not a difficult work. 
1. start a CompareFace service: https://github.com/exadel-inc/CompreFace#getting-started-with-compreface
- Download the archive from latest release: https://github.com/exadel-inc/CompreFace/releases
- Unzip the archive
- Open the terminal in this folder and run this command: docker-compose up -d
- Open the service in your browser: http://localhost:8000/login

2. use CompareFace python sdk to extract face embeddings: https://github.com/exadel-inc/compreface-python-sdk
looks like the script is simple, construct a recognizion 
```
DOMAIN: str = 'http://localhost'
PORT: str = '8000'
compre_face: CompreFace = CompreFace(DOMAIN, PORT)
recognition: RecognitionService = compre_face.init_face_recognition(API_KEY)
image_path: str = 'examples/common/jonathan-petit-unsplash.jpg'
response = recognition.recognize(image_path=image_path)
```
the response is a list of faces, each face has an embedding:
```
{
  "result" : [ {
    "age" : {
      "probability": 0.9308982491493225,
      "high": 32,
      "low": 25
    },
    "gender" : {
      "probability": 0.9898611307144165,
      "value": "female"
    },
    "mask" : {
      "probability": 0.9999470710754395,
      "value": "without_mask"
    },
    "embedding" : [ 9.424854069948196E-4, "...", -0.011415496468544006 ],
    "box" : {
      "probability" : 1.0,
      "x_max" : 1420,
      "y_max" : 1368,
      "x_min" : 548,
      "y_min" : 295
    },
    "landmarks" : [ [ 814, 713 ], [ 1104, 829 ], [ 832, 937 ], [ 704, 1030 ], [ 1017, 1133 ] ],
    "subjects" : [ {
      "similarity" : 0.97858,
      "subject" : "subject1"
    } ],
    "execution_time" : {
      "age" : 28.0,
      "gender" : 26.0,
      "detector" : 117.0,
      "calculator" : 45.0,
      "mask": 36.0
    }
  } ],
  "plugins_versions" : {
    "age" : "agegender.AgeDetector",
    "gender" : "agegender.GenderDetector",
    "detector" : "facenet.FaceDetector",
    "calculator" : "facenet.Calculator",
    "mask": "facemask.MaskDetector"
  }
}
```

3. insert the embeddings into milvus, or use the embedding to search in milvus
for example, I get a response from compare_face, insert the embeddings into milvus:
```
connections.connect(host=MILVUS_HOST, port=MILVUS_PORT)
response = recognition.recognize(image_path=image_path)
embeddings = []
for res in response["result"]:
    embeddings.insert(res.["embedding"])

data = [embeddings]
collection = Collection(collection_name)
collection.insert(data)

```

**Reply by martinenkoEduard** [reply]
Thanks!)

**Reply by martinenkoEduard** [reply]
Btw what is the best search algorithm to search similar faces? 
Is Euclidian distance a good choice?

**Reply by martinenkoEduard** [reply]
I get a following error:
    raise DataTypeNotSupportException(0, ExceptionsMessage.DataTypeNotSupport)
pymilvus.orm.exceptions.DataTypeNotSupportException: <DataTypeNotSupportException: (code=0, message=Data type is not support.)>

using following scheam

fields = [
    FieldSchema(name="pk", dtype=DataType.INT64, is_primary=True, auto_id=False),   
    FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=512)
]

Resolved it!

Just need to use mr = collection.insert([[0],[embedings]])

**Reply by yhmo** [reply]
> Btw what is the best search algorithm to search similar faces? Is Euclidian distance a good choice?

“search algorithm” is not a proper concept.
For ANN search, we have index algorithm and metric method.
Metric method is to determine the similarity of two vectors. Index algorithm is to accelerate search speed of ANN search.
For float vector, milvus has two kinds of metric method: L2 and IP. L2 is Euclidian distance, IP is inner product method. For normalized vectors, Euclidean distance (L2) is mathematically equivalent to inner product (IP).
